[Sol]Step02-2 - 카드 클래스 구현하기

Hansolkkim opened the pull request:
# 02-2) 카드 클래스 구현하기
## 작업 목록
- [x] Card 클래스 생성
  - [x] 해당 클래스의 정보를 String 타입으로 반환하는 `makeDescription()` 메서드 선언
- [x] 특정 카드 인스턴스를 생성해 출력
## 학습 키워드
`Type` , `enum`
## 고민과 해결
- 이 때까지 어떤 데이터의 타입을 큰 고민없이 결정했었는데, 이번 Step을 진행하면서 왜 이 타입을 결정해야하는지 좀 더 생각하는 시간을 가지게 되었습니다. 변수명을 선택하는 것도 어렵지만, 타입을 결정하는데도 어느정도 고민이 필요하다는걸 깨달았습니다.

Hansolkkim committed:
Fix: 코드 리뷰에 따른 코드 수정

Hansolkkim committed:
Docs: README.md에 Step1 진행사항, 추가학습거리, 코드리뷰 피드백 내용 추가

Hansolkkim committed:
Feat: 카드의 number, symbol을 프로퍼티로 갖는 Card class 선언

Hansolkkim committed:
Docs: 앱 아이콘 추가

Hansolkkim committed:
Feat: 임의의 Card 인스턴스를 생성하고 카드의 정보를 콘솔에 출력하는 코드 추가

godrm commented on PockerGameApp/PockerGameApp/Card.swift:
```diff
@@ -0,0 +1,48 @@
+        self.symbol = symbol
+    }
+    
+    func makeDescription() -> String {
```
description을 처리하는 것을 지원하기 위해서 CustomStringConvertible 프로토콜을 학습해보세요

Hansolkkim replied:
> description을 처리하는 것을 지원하기 위해서 CustomStringConvertible 프로토콜을 학습해보세요

description에 해당하는 값을 반환하는 메서드라고 단정지어서 생각해서, 공부했었던 프로토콜에 대해서 생각을 못 했네요,, 감사합니다!

godrm commented on PockerGameApp/PockerGameApp/Card.swift:
```diff
@@ -0,0 +1,48 @@
+            numberValue = String(number.rawValue)
+            
+        }
+        return "모양: \(symbol.rawValue), 숫자: \(numberValue)"
```
numberValue 처리와 symbol 모두 여기에 있는데, Number와 Symbol 타입을 선언했으니까 관련된 부분으로 옮겨보세요
마찬가지로 CustomStringConvertible을 사용해도 됩니다

Hansolkkim replied:
넵 수정 후 재PR 하도록 하겠습니다!

godrm commented on PockerGameApp/PockerGameApp/ViewController.swift:
```diff
@@ -9,24 +9,35 @@ import UIKit
+        let cardHeight = 1.27*Float(cardWidth)
+        let spacingFromTop = 47
+        for eachImageViewIndex in 0..<maxNumberOfCards {
+            let xPosition = CGFloat(5*(1+eachImageViewIndex) + cardWidth*eachImageViewIndex)
```
아직 5나 1 같은 숫자가 하드코딩되어 있습니다. 상수로 표현해도 좋을 것 같습니다

Hansolkkim replied:
아직 남아있는 숫자들이 있었네요, 수정하겠습니다

godrm commented on PockerGameApp/PockerGameApp/ViewController.swift:
```diff
@@ -9,24 +9,35 @@ import UIKit
+        let spacingFromTop = 47
+        for eachImageViewIndex in 0..<maxNumberOfCards {
+            let xPosition = CGFloat(5*(1+eachImageViewIndex) + cardWidth*eachImageViewIndex)
+            let currentFrame = CGRect(x: CGFloat(xPosition), y: CGFloat(spacingFromTop), width: CGFloat(cardWidth), height: CGFloat(cardHeight))
```
CGFloat로 타입 변환을 하기보다 선언할 때 타입을 지정해도 되지 않을까요? 어떤 차이가 있을까요?

Hansolkkim replied:
> CGFloat로 타입 변환을 하기보다 선언할 때 타입을 지정해도 되지 않을까요? 어떤 차이가 있을까요?

추후에 해당값을 바꾸려할 때 좀 더 직관적으로 바꿀 수 있다고 생각하여 선언 시 타입을 지정하지 않았는데, 큰 차이가 없는 것 같습니다. CGRect 초기화하는 부분이 너무 길어 가독성이 떨어지는 문제가 더 큰 것 같아 수정하도록 하겠습니다.

godrm commented on PockerGameApp/PockerGameApp/ViewController.swift:
```diff
@@ -9,24 +9,35 @@ import UIKit
+            let newImageView = UIImageView(frame: currentFrame)
+            newImageView.image = cardImage
+            self.view.addSubview(newImageView)
+            cards.append(UIImageView(frame: currentFrame))
```
self.view.addSubview()로도 추가를 하는데 cards.append()를 다시 하는 이유가 뭘까요?
안해도 상관없지 않을까요?

Hansolkkim replied:
> self.view.addSubview()로도 추가를 하는데 cards.append()를 다시 하는 이유가 뭘까요?
> 안해도 상관없지 않을까요?

나중에 각각의 카드에 해당하는 Image를 수정할 때, 배열로 접근하면 더 수월하게 바꿀 수도 있고, 변수로 지정해야 나중에 변경이 편하다고 생각되어 cards 배열에 넣는 방법을 사용했습니다. 공부했던 내용을 되짚어봤더니, cards라는 배열 없이도 self.view.subviews 배열로도 인덱스를 이용해 접근할 수 있네요. cards 배열을 삭제하도록 하겠습니다.

godrm reviewed (changes requested):
다음 부분 중에서 프로토콜 관련된 부분은 수정해서 다시 Push해보세요

Hansolkkim committed:
Fix: 코드 리뷰에 따른 코드 수정

Number, Symbol, Card가 CustomStringConvertible 프로토콜을 채택하도록 수정
그에 따른 printDescription(_:) 메서드 수정
숫자로 하드코딩된 값들을 상수로 선언하여 setImageView() 메서드 수정
불필요한 cards 배열 삭제

Hansolkkim commented:
수정했습니다.

Hansolkkim committed:
Fix: Symbol enum에 있는 기호들을 유니코드로 수정

godrm commented on PockerGameApp/PockerGameApp/Card.swift:
```diff
@@ -27,22 +46,11 @@ class Card {
+
+extension Card: CustomStringConvertible {
+    var description: String {
+        return symbol.description + number.description
```
문자열을 합성할 때는 `+` 방식보다 `"\(symbol)\(number)"` 형태로 Interpolation 방식을 추천합니다.

Hansolkkim replied:
수정해서 다음 PR에 추가하겠습니다. 감사합니다!

godrm commented on PockerGameApp/PockerGameApp/ViewController.swift:
```diff
@@ -8,35 +8,35 @@
-    func printCardDescription(_ card: Card) {
-        print(card.makeDescription())
+    func printDescription(of target: CustomStringConvertible) {
+        print(target.description)
```
참고로 print()할 때는 description을 붙이지 않아도 CustomStringConvertible를 구현했으면 description 값이 출력됩니다.

Hansolkkim replied:
감사합니다!
